File: backend/app/schemas/skill.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field, HttpUrl, field_validator

from app.models.skill import EntryType, SkillStatus, VisibilityEnum
from app.services.scanner import FileManifestEntry


_GITHUB_URL_RE = re.compile(r"^https://github\.com/[^/]+/[^/]+$")


class LabelOut(BaseModel):
    name: str
    usage_count: int
    applied_by_me: bool = False


class AdminLabelOut(BaseModel):
    id: str
    name: str
    usage_count: int


class SkillCreate(BaseModel):
    repo_url: str
    skill_path: str = "/"
    name: Optional[str] = None
    description: Optional[str] = None
    compatible_platforms: List[str] = []
    keywords: List[str] = []  # deprecated: converted to labels on create
    license: Optional[str] = None
    version: Optional[str] = None
    uses_agent_gateway: bool = False
    entry_type: EntryType = EntryType.skill
    # local submission fields (source_type='local')
    source_type: str = "github"
    snapshotted_files: Dict[str, str] = Field(default_factory=dict)


class SkillUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    compatible_platforms: Optional[List[str]] = None
    license: Optional[str] = None
    version: Optional[str] = None
    uses_agent_gateway: Optional[bool] = None
    superseded_by_slug: Optional[str] = None
    changelog_note: Optional[str] = None
    forked_from_url: Optional[str] = None

    @field_validator("forked_from_url")
    @classmethod
    def validate_forked_from_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        from app.services.github import _normalize_github_url
        normalized = _normalize_github_url(v)
        if normalized and not _GITHUB_URL_RE.match(normalized):
            raise ValueError("forked_from_url must be a https://github.com/<owner>/<repo> URL")
        return normalized


class RateSkillIn(BaseModel):
    value: int = Field(..., ge=1, le=5)


class RateSkillOut(BaseModel):
    avg_rating: float
    rating_count: int
    my_rating: int


class SkillOut(BaseModel):
    id: str
    slug: str
    name: str
    repo_url: str
    skill_path: str
    entry_type: EntryType
    status: SkillStatus
    deactivation_reason: Optional[str]
    superseded_by_slug: Optional[str]
    description: Optional[str]
    readme_html: Optional[str]
    skill_md_raw: Optional[str]
    skill_md_filename: Optional[str]
    readme_raw: Optional[str]
    compatible_platforms: List[str]
    license: Optional[str]
    version: Optional[str]
    github_stars: Optional[int]
    last_commit_at: Optional[datetime]
    readme_fetched_at: Optional[datetime]
    uses_agent_gateway: bool
    visibility: VisibilityEnum
    forked_from_url: Optional[str]
    # plugin.json metadata
    agent_count: int = 0
    agent_names: List[str] = []
    has_mcp_server: bool = False
    has_scripts: bool = False
    plugin_author: Optional[str] = None
    # source tracking
    source_type: str = "github"
    # file manifest
    file_manifest: List[FileManifestEntry] = []
    manifest_truncated: bool = False
    # version pinning (#017)
    pinned_commit_sha: Optional[str] = None
    pinned_ref: Optional[str] = None
    upstream_sha: Optional[str] = None
    update_available: bool = False

    submitter_id: str
    submitted_at: datetime
    updated_at: datetime
    avg_rating: float
    rating_count: int
    flag_count: int
    labels: List[LabelOut] = []
    my_rating: Optional[int] = None
    my_flag: Optional[object] = None  # FlagOut | None — avoid circular import; typed at route layer

    model_config = {"from_attributes": True}


class SkillListOut(BaseModel):
    id: str
    slug: str
    name: str
    entry_type: EntryType
    status: SkillStatus
    description: Optional[str]
    compatible_platforms: List[str]
    github_stars: Optional[int]
    avg_rating: float
    rating_count: int
    flag_count: int
    visibility: VisibilityEnum
    forked_from_url: Optional[str]
    # plugin.json metadata
    agent_count: int = 0
    agent_names: List[str] = []
    has_mcp_server: bool = False
    has_scripts: bool = False
    plugin_author: Optional[str] = None
    # source tracking
    source_type: str = "github"
    # version pinning (#017)
    update_available: bool = False

    submitter_id: str
    submitted_at: datetime
    updated_at: datetime
    labels: List[LabelOut] = []

    model_config = {"from_attributes": True}


class PaginatedSkills(BaseModel):
    items: List[SkillListOut]
    total: int
    page: int
    page_size: int


class RevisionOut(BaseModel):
    revision_number: int
    actor_id: str
    action: str
    changelog_note: Optional[str]
    created_at: datetime
    snapshot: dict

    model_config = {"from_attributes": True}


class GitHubPreviewOut(BaseModel):
    name: str
    description: Optional[str]
    stars: int
    license: Optional[str]
    last_commit_at: Optional[datetime]
    visibility: VisibilityEnum


class GitHubRefOut(BaseModel):
    owner: str
    repo: str
    branch: Optional[str]
    path: str


class SkillScanSnapshotOut(BaseModel):
    ref: GitHubRefOut
    name: Optional[str]
    description: Optional[str]
    compatible_platforms: List[str]
    version: Optional[str]
    license: Optional[str]
    readme_html: Optional[str]
    stars: int
    last_commit_at: Optional[datetime]
    visibility: VisibilityEnum
    forked_from_url: Optional[str]
    fetched_at: datetime
    no_skill_files: bool
    existing_slug: Optional[str]
    # plugin.json fields
    agent_count: int = 0
    agent_names: List[str] = []
    has_mcp_server: bool = False
    has_scripts: bool = False
    plugin_author: Optional[str] = None
    keywords: List[str] = []
    # file manifest
    file_manifest: List[FileManifestEntry] = []
    manifest_truncated: bool = False


class DiscoverOut(BaseModel):
    skills: List[SkillScanSnapshotOut]
    tree_truncated: bool = False
    capped: bool = False


class SkillSummaryOut(BaseModel):
    """Slim skill record for the /api/skills/summary endpoint (no readme_html)."""
    slug: str
    name: str
    description: Optional[str]
    repo_url: str
    skill_path: str
    entry_type: str
    compatible_platforms: List[str]
    version: Optional[str] = None
    avg_rating: float
    rating_count: int
    labels: List[str] = []

    model_config = {"from_attributes": True}

File: frontend/types/skill.ts
export type EntryType = "skill" | "marketplace_ref";
export type SkillStatus = "active" | "deactivated";
export type VisibilityType = "public" | "internal" | "private";

export interface LabelOut {
  name: string;
  usage_count: number;
  applied_by_me: boolean;
}

export interface AdminLabelOut {
  id: string;
  name: string;
  usage_count: number;
}

export interface Skill {
  id: string;
  slug: string;
  name: string;
  repo_url: string;
  skill_path: string;
  entry_type: EntryType;
  status: SkillStatus;
  deactivation_reason: string | null;
  superseded_by_slug: string | null;
  description: string | null;
  readme_html: string | null;
  skill_md_raw: string | null;
  skill_md_filename: string | null;
  readme_raw: string | null;
  compatible_platforms: string[];
  license: string | null;
  version: string | null;
  github_stars: number | null;
  last_commit_at: string | null;
  submitter_id: string;
  submitted_at: string;
  updated_at: string;
  avg_rating: number;
  rating_count: number;
  flag_count: number;
  uses_agent_gateway: boolean;
  visibility: VisibilityType;
  forked_from_url: string | null;
  agent_count: number;
  agent_names: string[];
  has_mcp_server: boolean;
  has_scripts: boolean;
  plugin_author: string | null;
  file_manifest: FileManifestEntry[];
  manifest_truncated: boolean;
  // version pinning (#017)
  pinned_commit_sha: string | null;
  pinned_ref: string | null;
  upstream_sha: string | null;
  update_available: boolean;
  labels: LabelOut[];
  my_rating: number | null;
  my_flag: FlagOut | null;
}

export type FlagReason = "broken" | "stale" | "superseded" | "inappropriate" | "other";
export type FlagStatus = "active" | "resolved";

export interface FlagOut {
  reason: FlagReason;
  note: string | null;
  status: FlagStatus;
  created_at: string;
}

export interface FlagResponse {
  flag_count: number;
  my_flag: FlagOut | null;
}

export interface RetractResponse {
  flag_count: number;
}

export interface FlaggedSkillItem {
  skill_slug: string;
  skill_name: string;
  flag_count: number;
  flags: Array<{
    reason: FlagReason;
    note: string | null;
    status: FlagStatus;
    created_at: string;
    reporter_id: string;
  }>;
}

export interface PaginatedFlaggedSkills {
  items: FlaggedSkillItem[];
  total: number;
  page: number;
  pages: number;
}

export interface RateSkillOut {
  avg_rating: number;
  rating_count: number;
  my_rating: number;
}

export interface PaginatedSkills {
  items: Skill[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  next_cursor: string | null;
  prev_cursor: string | null;
  platform_counts?: Record<string, number>;
}

export interface SkillRevision {
  id: string;
  skill_id: string;
  revision_number: number;
  snapshot: Record<string, unknown>;
  actor_id: string;
  action: "create" | "edit" | "refetch" | "deactivate" | "reactivate" | "pin";
  changelog_note: string | null;
  created_at: string;
}

export type FieldDiff =
  | { field: string; type: "scalar"; old: string | number | null; new: string | number | null }
  | { field: string; type: "array"; added: string[]; removed: string[] }
  | { field: string; type: "readme_updated" };

export interface SkillCreate {
  repo_url: string;
  skill_path?: string;
  name?: string;
  description?: string;
  compatible_platforms?: string[];
  keywords?: string[];  // deprecated: converted to labels on create
  version?: string;
  license?: string;
}

export interface SkillUpdate {
  name?: string;
  description?: string;
  compatible_platforms?: string[];
  version?: string;
  license?: string;
  changelog_note?: string;
  forked_from_url?: string;
}

export interface User {
  user_id: string;
  is_admin: boolean;
}

export interface UserProfile {
  user_id: string;
  submitted_count: number;
  edited_count: number;
  install_count?: number; // only present for self or admin viewer
}

export interface InstallEvent {
  skill_slug: string;
  skill_name: string | null;
  installed_at: string;
  update_available: boolean;
  is_deleted: boolean;
}

export interface PaginatedInstalls {
  items: InstallEvent[];
  total: number;
  page: number;
  page_size: number;
}

export interface GitHubPreview {
  name: string;
  description: string | null;
  stars: number;
  license: string | null;
  last_commit_at: string | null;
  visibility: VisibilityType;
}

export type SortOption = "newest" | "highest_rated" | "most_rated" | "most_stars";

export interface SkillListParams {
  q?: string;
  labels?: string[];
  sort?: SortOption;
  page?: number;
  page_size?: number;
  forked_from?: string;
  visibility?: VisibilityType | "all";
  cursor?: string;
  platforms?: string[];
}

export interface GitHubRef {
  owner: string;
  repo: string;
  branch: string | null;
  path: string;
}

export interface FileManifestEntry {
  path: string;
  size_bytes: number;
  is_text: boolean;
  is_dir: boolean;
}

export interface SkillScanSnapshot {
  ref: GitHubRef;
  name: string | null;
  description: string | null;
  compatible_platforms: string[];
  version: string | null;
  license: string | null;
  readme_html: string | null;
  stars: number;
  last_commit_at: string | null;
  visibility: VisibilityType;
  forked_from_url: string | null;
  fetched_at: string;
  no_skill_files: boolean;
  existing_slug: string | null;
  agent_count: number;
  agent_names: string[];
  has_mcp_server: boolean;
  has_scripts: boolean;
  plugin_author: string | null;
  keywords: string[];
  file_manifest: FileManifestEntry[];
  manifest_truncated: boolean;
}

export interface DiscoverResult {
  skills: SkillScanSnapshot[];
  tree_truncated: boolean;
  capped: boolean;
}
